### PR TITLE
one operation unstake

### PR DESCRIPTION
**Pull request by matusvla**
- First version of possible fix of the rewards distribution issue
- Test fix
- Bonds adjusted to the offset approach
- JS adjustments
- JS test adjustment - not working
- Fixed delta calculation
- test "framework" in progress
- test "framework" in progress - mint
- test "framework" in progress - pool creations
- test "framework" partially finished
- test "framework" error handling improved
- test_runner externalized
- bond testing in progress
- bond testing
- stake and unstake allowed for an uncranked pool
- cleanup
- test that should succeed
- bonds included in the staking minimum
- unstakes under the minimum pool limit forbidden
- Refreshing the minimum stake amounts
- Optional bond account parameter added to the unstake request
- JS bindings for stake and unstake updated
- JS test fix in progress
- JS test passing (but some of the parts commented out)
- one operation unstake implemented

### CODE REVIEW

**Comment by matusvla** [comment]
dependent on #51 and #52

File: smart-contract/program/src/processor/claim_bond_rewards.rs [hunk 1]
```diff
@@ -152,11 +152,12 @@ pub fn process_claim_bond_rewards(
     )?
     // Multiply by the staker shares of the total pool
     .checked_mul(bond.total_staked as u128)
-    .map(|r| r >> 32)
+    .map(|r| ((r >> 31) + 1) >> 1)
```
Comment (martincik) [inline smart-contract/program/src/processor/claim_bond_rewards.rs:155]: 🚀

File: smart-contract/program/src/state.rs [hunk 2]
```diff
@@ -44,7 +44,7 @@ pub const MAX_UNSTAKE_REQUEST: usize = 10;
 /// Fees charged on staking instruction in % (i.e FEES = 1 <-> 1% fee charged)
 pub const FEES: u64 = 2;
 
-#[derive(BorshSerialize, BorshDeserialize, BorshSize, PartialEq, FromPrimitive, ToPrimitive)]
+#[derive(BorshSerialize, BorshDeserialize, BorshSize, PartialEq, FromPrimitive, ToPrimitive, Debug)]
```
Comment (martincik) [inline smart-contract/program/src/state.rs:47]: tbr

File: smart-contract/program/tests/unit.rs [hunk 3]
```diff
@@ -0,0 +1,187 @@
+use solana_test_framework::*;
+
+
+
+
```
Comment (martincik) [inline smart-contract/program/tests/unit.rs:5]: tbr

File: smart-contract/program/tests/full.rs [hunk 4]
```diff
@@ -0,0 +1,395 @@
+use solana_test_framework::*;
+
```
Comment (martincik) [inline smart-contract/program/tests/full.rs:2]: tbr

File: smart-contract/program/src/state.rs (hunk 2 above)
Comment (matusvla) [reply smart-contract/program/src/state.rs:47]: Needed for testing

**Pull request merged**